build_runner/test/server/hot_reload_client/module_test.dart

```diff
@@ -0,0 +1,249 @@
+
+void main() {
+  group('roughLibraryKeyDecode', () {
+    var data = [
+      {'module': 'web/main', 'key': 'main', 'result': 'web/main'},
+      {'module': 'web/main', 'key': 'other', 'result': 'web/other'},
+      {'module': 'packages/a/a', 'key': 'a', 'result': 'packages/a/a'},
+      {'module': 'packages/a/a', 'key': 'other', 'result': 'packages/a/other'},
+      {
+        'module': 'packages/a/a',
+        'key': 'with_underscore',
+        'result': 'packages/a/with_underscore'
+      },
+      {
+        'module': 'packages/a/a',
+        'key': 'with__subdir',
+        'result': 'packages/a/with/subdir'
+      },
+      {
+        'module': 'packages/a/a',
+        'key': r'with$46period',
+        'result': 'packages/a/with.period'
+      },
+      {
+        'module': 'packages/a/a',
+        'key': r'subdir__underscore_and$46period',
+        'result': 'packages/a/subdir/underscore_and.period'
+      },
+    ];
+    for (var item in data) {
+      test('should produce ${item['result']}', () {
+        expect(Module.roughLibraryKeyDecode(item['module'], item['key']),
+            item['result']);
+      });
+    }
+  });
+
+  group('instance methods', () {
+    Module module;
+
+    setUp(() {
+      var libs = <String, MockLibrary>{};
+      libs['1'] = named(MockLibrary(), name: '1');
+      libs['2'] = named(MockLibrary(), name: '2');
+      libs['3'] = named(MockLibrary(), name: '3');
+      module = Module(libs);
+    });
+
+    test('onDestroy should run on all and collect data', () {
+      when(module.libraries['1'].onDestroy()).thenReturn(1);
+      when(module.libraries['2'].onDestroy()).thenReturn('two');
+      expect(module.onDestroy(), {'1': 1, '2': 'two', '3': null});
+      verify(module.libraries['1'].onDestroy()).called(1);
+      verify(module.libraries['2'].onDestroy()).called(1);
+      verify(module.libraries['3'].onDestroy()).called(1);
+    });
+
+    group('onSelfUpdate', () {
+      test('onSelfUpdate returns true if all are true', () {
+        when(module.libraries['1'].onSelfUpdate(any)).thenReturn(true);
+        when(module.libraries['2'].onSelfUpdate(any)).thenReturn(true);
+        when(module.libraries['3'].onSelfUpdate(any)).thenReturn(true);
+        expect(module.onSelfUpdate({}), true);
+      });
+
+      test('onSelfUpdate returns null if any is null and has no false', () {
+        when(module.libraries['1'].onSelfUpdate(any)).thenReturn(true);
+        when(module.libraries['2'].onSelfUpdate(any)).thenReturn(true);
+        expect(module.onSelfUpdate({}), null);
+      });
+
+      test('onSelfUpdate returns false if any is false', () {
+        when(module.libraries['1'].onSelfUpdate(any)).thenReturn(true);
+        when(module.libraries['2'].onSelfUpdate(any)).thenReturn(true);
+        when(module.libraries['3'].onSelfUpdate(any)).thenReturn(false);
+        expect(module.onSelfUpdate({}), false);
+      });
+
+      test('onSelfUpdate should run all even if returns null', () {
```

Comment:
Ah ok this test pretty much answers my question above - seems like reasonable behavior actually.

Please add doc comments to the implementations about the expected behavior though.